src/NHibernate.Test/NHSpecificTest/GH1170/Model.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace NHibernate.Test.NHSpecificTest.GH1170
+{
+	public class Parent
+	{
+		public virtual ICollection<ChildComponent> ChildComponents { get; set; } = new List<ChildComponent>();
+		public virtual Guid Id { get; set; }
+	}
+
+	public class ChildComponent
+	{
+		public virtual bool SomeBool { get; set; }
+		public virtual string SomeString { get; set; }
+
+		protected bool Equals(ChildComponent other)
+		{
+			return SomeBool == other.SomeBool && string.Equals(SomeString, other.SomeString);
+		}
+
+		public override bool Equals(object obj)
```

Comment:
Not overriding it causes the set to always delete all its elements, causing the update test to be even more moot. (It is anyway somewhat moot, has with set, updates are handled by delete/insert.)

Reply:
I think this behaviour needs to be changed (in a separate PR).

Reply:
Do you mean, if the component reference has not changed, update it instead of deleting/inserting it? But the persistence context loaded state does not keep track of component references and so, does not allow to check for that condition.
Or do you mean we should somewhat "recycle" components to delete by updating them with data of components to add in order to replace as most as possible deletes/inserts by single updates? That would cause actual removals combined with adds to be replaced by updates, which could be quite unexpected for users, and breaking for custom auditing logic (like trigger based auditing).

Reply:
I'm talking about this behaviour: "Not overriding it causes the set to always delete all its elements" it shouldn't be like this.

Reply:
From [here](http://nhibernate.info/doc/nhibernate-reference/components.html#components-incollections):

> Note: if you define an `ISet` of composite elements, it is very important to implement `Equals()` and `GetHashCode()` correctly. 

It is in fact an error to map a set of component without a proper equal semantic. The set cannot enforce its semantic in such case: better use a bag instead.

Of course the side effect this error has about flushes, causing the whole set to be purged and recreated, could still be fixed. But that would be fixing a case resulting anyway from a bad mapping/implementation on the user side. Why not if this can be done in a simple way with very few impacts on normal uses, but otherwise, I do not think it should be fixed.

Fixing it will likely require to extract the set content as component values arrays, then searching for each array of values if we can find the same array in the loaded state. This looks to me as a non-negligible overhead, compared to just using the set `Contains` method. Moreover it would then ignore equals implementations having another semantic, which will be a possible breaking change.